storage/executable-store/src/lib.rs

```diff
@@ -0,0 +1,588 @@
+/// of the ExecutableStore throughout the system, as due to the current codebase structure
+/// constraints, different important steps are performed in different crates. Such
+/// defensive programming here ensures correctness (clearing the cache on any unexpected
+/// state updates) and observability to quickly identify any incorrect usage.
+///
+/// Initial state is Empty, and it is also possible to get to the Empty state by clearing
+/// the ExecutableStore at any time.
+///
+/// Execution in the system happens piece-wiese, where each piece contains a continuous
+/// sub-sequence of ledger of the ordered transactions. For example, in our codebase,
+/// pieces are called 'blocks' during the original execution / 'chunks' during state-sync.
+/// The only exception to this occurs while the cache in the Before(id) state is updated,
+/// adding and removing elements (potentially concurrently) in order to convert it to
+/// contain the executables at version corresponding to after the execution of the piece.
+/// When the update is done, it is the user's responsibility to mark the state as Updated.
+/// Currently, this happens after parallel execution in the block-executor crate.
+///
+/// After updating the ExecutableStore, it must be pruned to control the memory usage.
+/// Currently, this happens in aptos-vm based on a single threshold, clearing the store
+/// if the size exceeds the threshold. Then, it is the caller's responsibility to change
+/// the status to Pruned (ensuring we will never use the ExecutableStore without pruning).
+/// After pruning, the state may change to After(id), with the id of the latest piece.
+/// To go back to the Before(id') state, the parent of piece id' is conformed to be id,
+/// ensuring correctness. This currently occurs in execution/executor, as the information
+/// about pieces (ids and their relation to each other) is only available there.
+///
+/// The full state transition diagram is below:
+///
+/// Empty ----------> Updated(None)
+///                      |
+///                      |
+///                      ↓
+///                   Pruned(None)
+///                      |
+///                      |
+///                      ↓
+///   --------------- After(id) -------------> Before(id')
+///   |                                            |
+///   |                                            |
+///   ↓                                            |
+/// Empty ----------> Updated(Option<ID>) <---------
+///                     ...
+///                     ...
+/// Note: After transitioning to Updated from Empty, the exact state is Updated(None),
+/// while after transitioning from Before(id'), the state is Updated(Some(id')).
+/// Similar principle applies to the Pruned state. Keeping id' allows to also confirm
+/// the consecutive nature of piece ids when marking After state.
+///
+#[derive(Copy, Clone, PartialEq, Debug)]
+enum ExecutableStoreState<ID: Clone + Debug + PartialEq> {
```

Comment:
Do we have to be so vague about ID? Can't we make it just a u64 counter

Reply:
Originally I had it to HashValue (which is the block ID type), but then realized we will likely also enable it in StateSync, and not sure chunk ID is the same type (I think we'd have to match versions there), so that's why I turned it into generic. Happy to bring it back to a concrete type for now, but would have to be HashValue, lmk.